utils/frame/try-runtime/cli/src/commands/follow_chain.rs

```diff
@@ -42,51 +46,177 @@ pub struct FollowChainCmd {
+	uri: String,
+	_phantom: PhantomData<Block>,
+}
+
+#[async_trait]
+impl<Block: BlockT> HeaderProvider<Block> for RpcHeaderProvider<Block>
+where
+	Block::Header: DeserializeOwned,
+{
+	async fn get_header(&mut self, hash: Block::Hash) -> Block::Header {
+		rpc_api::get_header::<Block, _>(&self.uri, hash).await.unwrap()
+	}
+}
+
+/// Abstraction over RPC subscription for finalized headers.
+#[async_trait]
+trait HeaderSubscription<Block: BlockT>
+where
+	Block::Header: HeaderT,
+{
+	/// Await for the next finalized header from the subscription.
+	///
+	/// Returns `None` if either the subscription has been closed or there was an error when reading
+	/// an object from the client.
+	async fn next_header(&mut self) -> Option<Block::Header>;
+}
+
+#[async_trait]
+impl<Block: BlockT> HeaderSubscription<Block> for Subscription<Block::Header>
+where
+	Block::Header: DeserializeOwned,
+{
+	async fn next_header(&mut self) -> Option<Block::Header> {
+		match self.next().await {
+			Some(Ok(header)) => Some(header),
+			None => {
+				log::warn!("subscription closed");
+				None
+			},
+			Some(Err(why)) => {
+				log::warn!("subscription returned error: {:?}. Probably decoding has failed.", why);
+				None
+			},
+		}
+	}
+}
+
+/// Stream of all finalized headers.
+///
+/// Returned headers are guaranteed to be ordered. There are no missing headers (even if some of
+/// them lack justification).
+struct FinalizedHeaders<Block: BlockT, HP: HeaderProvider<Block>, HS: HeaderSubscription<Block>> {
+	header_provider: HP,
+	subscription: HS,
+	fetched_headers: VecDeque<Block::Header>,
+	last_returned: Option<<Block::Header as HeaderT>::Number>,
+}
+
+impl<Block: BlockT, HP: HeaderProvider<Block>, HS: HeaderSubscription<Block>>
+	FinalizedHeaders<Block, HP, HS>
+where
+	<Block as BlockT>::Header: DeserializeOwned,
+{
+	pub fn new(header_provider: HP, subscription: HS) -> Self {
+		Self {
+			header_provider,
+			subscription,
+			fetched_headers: VecDeque::new(),
+			last_returned: None,
+		}
+	}
+
+	/// Reads next finalized header from the subscription. If some headers (without justification)
+	/// have been skipped, fetches them as well.
+	///
+	/// All fetched headers are stored in `self.fetched_headers`.
+	async fn fetch(&mut self) {
+		let last_finalized = match self.subscription.next_header().await {
+			Some(header) => header,
+			None => return,
+		};
+
+		self.fetched_headers.push_front(last_finalized.clone());
+
+		let current_height = last_finalized.number();
+		let parent_height = current_height.sub(One::one());
+		let last_height = self.last_returned.unwrap_or(parent_height);
+
+		let mut parent_hash = last_finalized.parent_hash().clone();
+		for _ in 0u32..(parent_height.saturating_sub(last_height).try_into().unwrap_or_default()) {
+			let parent_header = self.header_provider.get_header(parent_hash).await;
+			self.fetched_headers.push_front(parent_header.clone());
+			parent_hash = *parent_header.parent_hash();
+		}
```

Comment:
if you change `last_returned` to contain the last hash returned, you can simplify this code to something like: 

```
let mut last_finalized_parent = last_finalized.parent_hash();
while last_finalized_parent != last_returned {
   // push header for last_finalized_parent
   // update last_finalized_parent
}
```